### PR TITLE
Test layers list model and view

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR adds comprehensive testing to our layers list model and view. It also fixes a couple minor issues encountered along the way with no changes to functionality.

It also exposed on bug, right now calling `layers.reverse()` doesn't seem to trigger the `reorder` event and so the layers list layout in the view doesn't get updated. Is there a way to make this happen @kne42?

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# How has this been tested?

- [x] adds `napari/components/_layers_list/tests/test_layers_list.py` and `napari/components/_layers_list/tests/test_layers_list_view.py`

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: napari/components/_layers_list/tests/test_layers_list.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+    """
+    layers = LayersList()
+    layer_a = Image(np.random.random((10, 10)))
+    layer_a.name = 'image_a'
+    layer_b = Image(np.random.random((15, 15)))
+    layer_b.name = 'image_b'
+    layers.append(layer_a)
+    layers.append(layer_b)
+
+    assert layers.index(layer_a) == 0
```
Comment (kne42) [inline napari/components/_layers_list/tests/test_layers_list.py:73]: you should also test with `'image_a'`

File: napari/components/_layers_list/tests/test_layers_list.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+    layers.append(layer_b)
+    layers.append(layer_c)
+
+    # Rearrange layers by tuple
+    layers[:] = layers[(1, 0, 2)]
+    assert [l for l in layers] == [layer_b, layer_a, layer_c]
+
+    # Swap layers by name
+    layers['image_b', 'image_c'] = layers['image_c', 'image_b']
+    assert [l for l in layers] == [layer_c, layer_a, layer_b]
```
Comment (kne42) [inline napari/components/_layers_list/tests/test_layers_list.py]:
```suggestion
    assert list(layers) == [layer_c, layer_a, layer_b]
```
(and elsewhere as applicable)

File: napari/components/_layers_list/tests/test_layers_list.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+    assert layers.index(layer_a) == 0
+
+
+def test_reordering():
+    """
+    Test indexing into a LayerList by name
+    """
+    layers = LayersList()
+    layer_a = Image(np.random.random((10, 10)))
+    layer_a.name = 'image_a'
```
Comment (kne42) [inline napari/components/_layers_list/tests/test_layers_list.py]: would probably be easier to read if you used the `name` kwarg above
(and elsewhere as applicable)

File: napari/components/_layers_list/tests/test_layers_list_view.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+def check_layout_layers(layout, layers):
+    """
+    Check the layer widget order matches the layers order in the layout
+
+    Parameters
+    ----------
+    layout : QLayout
+        Layout to test
+    layers : napar.components.LayersList
```
Comment (kne42) [inline napari/components/_layers_list/tests/test_layers_list_view.py]:
```suggestion
    layers : napari.components.LayersList
```

File: napari/components/_layers_list/view.py [hunk 5]
```diff
@@ -211,10 +211,9 @@ def dropEvent(self, event):
         layer_name = event.mimeData().text()
         index = self.layers.index(layer_name)
         if index != insert and index + 1 != insert:
-            if not self.layers[index].selected:
-                self.layers.unselect_all()
-                self.layers[index].selected = True
-            self.layers._move_layers(index, insert)
+            if insert >= index:
+                insert = insert - 1
```
Comment (kne42) [inline napari/components/_layers_list/view.py]:
```suggestion
                insert -= 1
```

**Review by kne42 — APPROVED** [review]
looks great @sofroniewn! so glad you're doing this :)

left a few minor comments on syntax/styling; conditionally approved once resolved

File: napari/components/_layers_list/tests/test_layers_list.py (hunk 1 above)
Comment (sofroniewn) [reply napari/components/_layers_list/tests/test_layers_list.py:73]: added, also added for `layer_b` and `"image_b"`

File: napari/components/_layers_list/tests/test_layers_list.py (hunk 2 above)
Comment (sofroniewn) [reply napari/components/_layers_list/tests/test_layers_list.py]: fixed here and elsewhere

File: napari/components/_layers_list/tests/test_layers_list.py (hunk 3 above)
Comment (sofroniewn) [reply napari/components/_layers_list/tests/test_layers_list.py]: changed

File: napari/components/_layers_list/tests/test_layers_list_view.py (hunk 4 above)
Comment (sofroniewn) [reply napari/components/_layers_list/tests/test_layers_list_view.py]: fixed

File: napari/components/_layers_list/view.py (hunk 5 above)
Comment (sofroniewn) [reply napari/components/_layers_list/view.py]: changed

**Pull request merged**